### PR TITLE
conditionally show cartoon network message on congrats page

### PR DESCRIPTION
**Pull request by davidsbailey**
When a user goes to https://studio.code.org/s/gumball/reset/lang/ar?partner=CN and then completes the gumball tutorial, we will now show a message linking back to cartoon network in arabic or english on the congrats page. We will not show the message if the same user completes a different tutorial, or closes the browser and then completes the gumball tutorial again, unless they access it using the same link.
## screenshots

english
screen shot 2015-12-03 at 3 01 28 pm

arabic
screen shot 2015-12-03 at 3 01 05 pm
## manual verification steps
1. CN message shows in Arabic when starting from CN link then going to congrats page
   - studio.code.org/s/gumball/reset/lang/ar?partner=CN
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - click "I have finished the Hour of Code"
   - --> Congrats page shows message in Arabic
2. CN message shows in English when changing language to English
   - studio.code.org/s/gumball/reset/lang/ar?partner=CN
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - change language to English
   - click "I have finished the Hour of Code"
   - --> Congrats page shows message in English
3. CN message does not show after completing another tutorial
   - studio.code.org/s/gumball/reset/lang/ar?partner=CN
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - studio.code.org/s/frozen/reset
   - get redirected to studio.code.org/s/frozen/stage/1/puzzle/1 in Arabic
   - click "I have finished the Hour of Code"
   - --> Congrats page does not show CN message
4. closing browser clears cookie even when not incognito
   - open studio.code.org/s/gumball/reset/lang/ar?partner=CN (not in incognito)
   - close browser
   - open studio.code.org/s/gumball/reset/
   - click "I have finished the Hour of Code"
   - --> Congrats page does not show CN message

### CODE REVIEW

**Comment by davidsbailey** [comment]
@laurelfan can you please review? @tanya please make sure this matches the behavior you're expecting. @wjordan , I'm adding a new cookie here. Open to other approaches if you dislike this.

File: cookbooks/cdo-varnish/libraries/http_cache.rb [hunk 1]
```diff
@@ -22,6 +22,7 @@ def self.config(env)
     whitelisted_cookies = [
       'hour_of_code',
       'language_',
+      'partner',
```
Comment (laurelfan) [inline cookbooks/cdo-varnish/libraries/http_cache.rb]: do you need to bump the version of the cdo-varnish cookbook?

File: dashboard/app/controllers/home_controller.rb [hunk 2]
```diff
@@ -10,6 +10,7 @@ class HomeController < ApplicationController
 
   def set_locale
     set_locale_cookie(params[:locale]) if params[:locale]
+    set_partner_cookie(params['partner']) if params['partner']
```
Comment (laurelfan) [inline dashboard/app/controllers/home_controller.rb]: you can use the symbol :partner here (params is an indifferent hash so string/symbol means the same thing)

**Comment by laurelfan** [comment]
LGTM, except ask @wjordan about the cdo-varnish cookbook version

**Comment by davidsbailey** [comment]
@philbogle this change is requested by our Cartoon Network partner and is needed for HoC. Can you please sign off? 

Thanks for the review @laurelfan , I've addressed your comments.

File: cookbooks/cdo-varnish/libraries/http_cache.rb (hunk 1 above)
Comment (tanyaparker) [inline cookbooks/cdo-varnish/libraries/http_cache.rb]: I'm about to bump the version for another http_cache change with https://github.com/code-dot-org/code-dot-org/pull/5876

@wjordan if I merge in my PR after Dave's will that take care of both changes then? So we don't have to bump it twice?

Comment (tanyaparker) [inline cookbooks/cdo-varnish/libraries/http_cache.rb]: Oh I see, @davidsbailey already made the version change. I guess to confirm, that once his PR is merged in then my change will also take effect?

**Comment by tanyaparker** [comment]
Yes, this matches the expected behavior. And the view looks good. Thanks, Dave!

**Comment by philbogle** [comment]
Sorry, this can't be submitted in its current form.  

At this late date, we can't accept PRs that change our cache configuration.  Adding another cache key reduces cachability and runs the risk of outrgiht caching regressions, and greatly slows deploys.

 Instead, set client state in localStorage or  sessionStorage and make decisions about whether show or hide content client side, or make Ajax requests.  I will think about this some more later tonight.

File: pegasus/sites.v3/code.org/public/congrats.haml [hunk 3]
```diff
@@ -49,10 +49,15 @@ max_age: 60
   -else
     %h1=I18n.t :congratulations_on_completing_hoc
 
-  -if request.locale == 'ar-SA'
-    %h2
-      أنهيت ساعة البرمجة ! العودة إلى
-      %a{:href=>"http://click-and-share.com/hour_of_code/data-capture/"} كرتون نتورك
+  -if params[:partner] == 'CN' && cert_script == 'gumball'
```
Comment (philbogle) [inline pegasus/sites.v3/code.org/public/congrats.haml]: You can rewrite this as a client-side code that looks at the cookie and decides to show or hide a div with the cartoon network link.

**Comment by davidsbailey** [comment]
Thank you Phil, good point about the caching change being risky. I will try to think of another way to do this without adding a new cookie.

**Comment by philbogle** [comment]
Note that I think it's fine to write the cookie (you'll be on a page that isn't stripping it), and that you can read the cookie from the javascript even on a cached page.  So then the Javascript can either pass the cookie along with the request or conditionally show a div depending on the value of the cookie.

**Comment by davidsbailey** [comment]
It appears we only allow language cookies through to `/congrats`, see https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-varnish/libraries/http_cache.rb#L78 . I've updated `/congrats` to read the cookie directly, and added `partner` to the cookie whitelist for `/congrats`. Does this help?

I have another approach in mind which piggy-backs off of the "company" param already used by the congrats page. I will send it out in the morning.

**Comment by davidsbailey** [comment]
Safer approach here: https://github.com/code-dot-org/code-dot-org/pull/5909 @tanyaparker @philbogle @laurelfan can you please take a look?

**Comment by davidsbailey** [comment]
closing without merging.

**Pull request closed without merging**